rust_src/src/windows.rs

```diff
@@ -1536,4 +1592,132 @@ pub fn window_body_width_lisp(window: LispWindowLiveOrSelected, pixelwise: bool)
     window.body_width(pixelwise).into()
 }
 
+/// Return pixel dimensions of WINDOW's lines.
+/// The return value is a list of the x- and y-coordinates of the lower
+/// right corner of the last character of each line.  Return nil if the
+/// current glyph matrix of WINDOW is not up-to-date.
+///
+/// Optional argument WINDOW specifies the window whose lines' dimensions
+/// shall be returned.  Nil or omitted means to return the dimensions for
+/// the selected window.
+///
+/// FIRST, if non-nil, specifies the index of the first line whose
+/// dimensions shall be returned.  If FIRST is nil and BODY is non-nil,
+/// start with the first text line of WINDOW.  Otherwise, start with the
+/// first line of WINDOW.
+///
+/// LAST, if non-nil, specifies the last line whose dimensions shall be
+/// returned.  If LAST is nil and BODY is non-nil, the last line is the last
+/// line of the body (text area) of WINDOW.  Otherwise, last is the last
+/// line of WINDOW.
+///
+/// INVERSE, if nil, means that the y-pixel value returned for a specific
+/// line specifies the distance in pixels from the left edge (body edge if
+/// BODY is non-nil) of WINDOW to the right edge of the last glyph of that
+/// line.  INVERSE non-nil means that the y-pixel value returned for a
+/// specific line specifies the distance in pixels from the right edge of
+/// the last glyph of that line to the right edge (body edge if BODY is
+/// non-nil) of WINDOW.
+///
+/// LEFT non-nil means to return the x- and y-coordinates of the lower left
+/// corner of the leftmost character on each line.  This is the value that
+/// should be used for buffers that mostly display text from right to left.
+///
+/// If LEFT is non-nil and INVERSE is nil, this means that the y-pixel value
+/// returned for a specific line specifies the distance in pixels from the
+/// left edge of the last (leftmost) glyph of that line to the right edge
+/// (body edge if BODY is non-nil) of WINDOW.  If LEFT and INVERSE are both
+/// non-nil, the y-pixel value returned for a specific line specifies the
+/// distance in pixels from the left edge (body edge if BODY is non-nil) of
+/// WINDOW to the left edge of the last (leftmost) glyph of that line.
+///
+/// Normally, the value of this function is not available while Emacs is
+/// busy, for example, when processing a command.  It should be retrievable
+/// though when run from an idle timer with a delay of zero seconds.
+#[lisp_fn(min = "0")]
+pub fn window_lines_pixel_dimensions(
+    window: LispWindowLiveOrSelected,
+    first: Option<LispNumber>,
+    last: Option<LispNumber>,
+    body: bool,
+    inverse: bool,
+    left: bool,
+) -> LispObject {
+    let window: LispWindowRef = window.into();
+    let buffer: LispBufferRef = window.contents.into();
```

Comment:
This is good. The order we check types is important. If we checked the buffer before the window the wrong error can be raised.